### PR TITLE
feat(python): expose storage options

### LINKED ISSUE #1930: Add `storage_options` parameter to `lance.dataset`
Currently lance.dataset can read from any Azure/S3/GCP uri, provided the environment variables allow it to authenticate with the cloud provider.

However, it'd be good to allow `lance.dataset(uri, storage_options)`, where `storage_options` is a `dict` as seen here:

```python

import polars as pl

source = "s3://bucket/*.parquet"

storage_options = {
    "aws_access_key_id": "<secret>",
    "aws_secret_access_key": "<secret>",
    "aws_region": "us-east-1",
}

df = pl.scan_parquet(source, storage_options=storage_options)

```

This is important when you need to authenticate a read operation to a bucket, but you don't want to change the env variable.

### PR DESCRIPTION
**Pull request by wjones127**
This allows users to pass object storage configuration in Lance in Python. Previously, users could only configure these settings through environment variables.

Closes #1930

### CODE REVIEW

**Comment by wjones127** [comment]
Manually tested against S3 with script:

```python
import pyarrow as pa
import lance
import os

data = pa.table({
    'x': pa.array([1, 2, 3, 4, 5]),
})

base_uri = 's3://some-bucket'
uri1 = os.path.join(base_uri, 'test1')
uri2 = os.path.join(base_uri, 'test2')

ds1 = lance.write_dataset(data, uri1, storage_options=dict(aws_region='us-east-1'))

ds2 = lance.write_dataset(
    data,
    uri2,
    storage_options=dict(
        aws_region='us-east-1',
        aws_server_side_encryption="aws:kms",
        aws_sse_kms_key_id='some-key-id',
    )
)

ds2.to_table()

ds3 = lance.dataset(uri2, storage_options=dict(aws_region='us-east-1'))
ds3.to_table()
```

File: docs/read_and_write.rst [hunk 1]
```diff
@@ -482,37 +482,129 @@ and Google Cloud Storage. Which object store to use is determined by the URI sch
...
+     - Timeout for the entire request, from connection until the response body
+       has finished. Default, ``30s``.
+   * - ``user_agent``
+     - User agent string to use in requests.
+   * - ``proxy_url``
+     - URL of a proxy server to use for requests. Default, ``None``.
+   * - ``proxy_ca_certificate``
+     - PEM-formatted CA certificate for proxy connections
+   * - ``proxy_excludes``
+     - List of hosts that bypass proxy
```
Comment (westonpace) [inline docs/read_and_write.rst:534]: What's the format?  Comma separated?

File: docs/read_and_write.rst [hunk 2]
```diff
@@ -482,37 +482,129 @@ and Google Cloud Storage. Which object store to use is determined by the URI sch
...
 
 
 S3 Configuration
 ~~~~~~~~~~~~~~~~
 
-To configure credentials for AWS S3, you can use the ``AWS_ACCESS_KEY_ID``,
-``AWS_SECRET_ACCESS_KEY``, and ``AWS_SESSION_TOKEN`` environment variables.
+.. code-block:: python
```
Comment (westonpace) [inline docs/read_and_write.rst:540]:
```suggestion
S3 has several authorization options that generally have to be specified.  If these are not specified here then the defaults will be loaded from the user's profile file (e.g. ~/.aws/config)

.. code-block:: python
```

Maybe still have some kind of initial sentence describing the section?

File: docs/read_and_write.rst [hunk 3]
```diff
@@ -482,37 +482,129 @@ and Google Cloud Storage. Which object store to use is determined by the URI sch
...
+   * - ``aws_endpoint`` / ``endpoint``
+     - The endpoint to use for S3-compatible stores.
+   * - ``aws_virtual_hosted_style_request`` / ``virtual_hosted_style_request``
+     - Whether to use virtual hosted-style requests, where bucket name is part
+       of the endpoint. Meant to be used with ``aws_endpoint``. Default, ``False``.
+   * - ``aws_s3_express`` / ``s3_express``
+     - Whether to use S3 Express One Zone endpoints. Default, ``False``. See more
+       details below.
+   * - ``aws_server_side_encryption``
+     - The server-side encryption algorithm to use. Must be one of "AES256",
```
Comment (westonpace) [inline docs/read_and_write.rst:581]:
```suggestion
     - The server-side encryption algorithm to use. Must be one of `"AES256"`,
```

File: docs/read_and_write.rst [hunk 4]
```diff
@@ -616,23 +709,94 @@ For details on how this feature works, please see :ref:`external-manifest-store`
 Google Cloud Storage Configuration
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-GCS credentials are configured by setting the ``GOOGLE_SERVICE_ACCOUNT`` environment
```
Comment (westonpace) [inline docs/read_and_write.rst:619]: Can we keep the first sentence here?

File: docs/read_and_write.rst [hunk 5]
```diff
@@ -616,23 +709,94 @@ For details on how this feature works, please see :ref:`external-manifest-store`
...
+   * - ``google_service_account_key`` / ``service_account_key``
+     - The serialized service account key.
+   * - ``google_application_credentials`` / ``application_credentials``
+     - Path to the application credentials.
+
+
 Azure Blob Storage Configuration
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Azure Blob Storage credentials can be configured by setting the ``AZURE_STORAGE_ACCOUNT_NAME``
```
Comment (westonpace) [inline docs/read_and_write.rst:634]: Same here, can we keep the first sentence?

**Review by westonpace — APPROVED** [review]
This is great, I have a few minor suggestions but this has been needed for a while, thank you.

**Pull request merged**